### DISCUSSION TITLE
Using persistant Zustand store within Next.js hydrating Zustand before SSR

### DISCUSSION THREAD

**Discussion by christianjuth** [pmndrs/zustand#2788] [Ideas]
**TLDR; I think Zustand, when persisting within a Next.js app should let you hydrate Zustand server side and client side**

I had the idea to persist my Zustand store using cookies. This means I can read/write to the cookie client side, and Server Side Rendered (SSR) my Next.js site with Zustand already hydrated, without worrying about a SSR hydration html mismatch.

I created a demo that allows you to select a sort preference for a list. The preference is stored and persisted via Zustand, and the list is SSR with the list already sorted.

* [Try the demo on vercel.](https://zustand-persist-react-hydration.vercel.app/)
* [See the code.](https://github.com/christianjuth/zustand-persist-react-hydration)
  * The readme explains how Zustand is setup
  
 **I'm not sure if I'm doing a good job explaining, but I would love to get your thoughts:**
 * Is this a use case that interests you?
 * Is there a better way to do this?
 * Anything I can clarify better?

**Comment by dbritto-dev** [comment]
@christianjuth you can't hydrate on server side only on client side. I encourage you to read our SRR and Hydration guide.

**Reply by christianjuth** [reply]
@dbritto-dev check out my example! 

**I successfully was able to hydrate server side, it's just a bit hacky.**

Basically, what makes this possible is storing state in cookies since the cookie can be read server/client side. I'm able to hydrate zustand server side and client side with the same persisted value, preventing any Next.js hydration errors.

**Reply by dbritto-dev** [reply]
@christianjuth that implementation is flaky, you shouldn't use global stores with `Nextjs`, we talked about that in our `Nextjs` guide. At some point you would ended up facing hydration errors and messing up everything. Stores should be created per request, keep that in mind.

**Reply by christianjuth** [reply]
@dbritto-dev out of curiosity, what makes this flaky? Maybe it's not possible to guarantee that my hydration function is called before the store is used? I would have to read up on if Next.js is guaranteed to render `layout` before `page`, but I attempted to make hydration the first thing that gets called.

I did look at your SRR and Hydration guide, but I didn't see an obvious solution to what I'm trying to accomplish. But I also tend to use Next.js's default server without implementing my own express.js server to render React. So maybe I'm missing something, or maybe this isn't possible to do reliably?

If it was possible, I think this would be an awesome way to create persistent stores that play nicely with both Server Components and Client Components.

Appreciate you taking the time to respond to me!

**Reply by dbritto-dev** [reply]
@christianjuth sure, you need to keep in mind that stores should be create per request in order to keep them isolated between multiple users/sessions.

**Reply by christianjuth** [reply]
@dbritto-dev Next.js should handle this automatically. 

> Good to know: cookies() is a [Dynamic Function](https://nextjs.org/docs/app/building-your-application/rendering/server-components#server-rendering-strategies#dynamic-functions) whose returned values cannot be known ahead of time. Using it in a layout or page will opt a route into [dynamic rendering](https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-rendering) at request time.
> [(source)](https://nextjs.org/docs/app/api-reference/functions/cookies)

Durning SSR I'm using `import { cookies } from "next/headers"` to hydrate Zustand. When Next.js sees this function called in a page/layout, the page automatically becomes dynamic, which means it's re-rendered each request, and has access to that request's cookies.

So I think users/session isolation shouldn't be an issue here?

**Reply by dbritto-dev** [reply]
@christianjuth `dynamic rendering` only means that you rendering at request time but that doesn't mean that you are isolating stores between requests

**Reply by dbritto-dev** [reply]
seems you are trying to share data across requests through nextjs

**Reply by dbritto-dev** [reply]
you can even have the same recommendations using `redux` -> https://redux.js.org/usage/nextjs

**Reply by christianjuth** [reply]
@dbritto-dev oh that's an interesting point.

I could restructure my React app to prevent that by creating the store per request, and passing it down via React Context.

I'm not married to my specific solution, but I still think having a way to hydrate the store for SSR should be possible. And I think with a little more modification, I could make my solution work reliably?

Am I on the right track at least? Is this a bad idea? I could make this a library that hooks into Zustand, if you think this could be useful for others?
